service/operator/operator.go

```diff
@@ -0,0 +1,137 @@
+package operator
+
+import (
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/cenk/backoff"
+	apismetav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/pkg/api/v1"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/giantswarm/operatorkit/framework"
+)
+
+type Config struct {
+	BackOff   backoff.BackOff
+	Framework *framework.Framework
+	K8sClient kubernetes.Interface
+	Logger    micrologger.Logger
+
+	ResyncPeriod time.Duration
+}
+
+func DefaultConfig() Config {
+	return Config{
+		BackOff:   nil,
+		Framework: nil,
+		K8sClient: nil,
+		Logger:    nil,
+
+		ResyncPeriod: time.Duration(0),
+	}
+}
+
+type Operator struct {
+	// Dependencies.
+	backOff   backoff.BackOff
+	framework *framework.Framework
+	k8sClient kubernetes.Interface
+	logger    micrologger.Logger
+
+	// Internals.
+	bootOnce     sync.Once
+	mutex        sync.Mutex
+	resyncPeriod time.Duration
+}
+
+func New(config Config) (*Operator, error) {
+	if config.BackOff == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.BackOff must not be empty")
+	}
+	if config.K8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.K8sClient must not be empty")
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.Logger must not be empty")
+	}
+	if config.Framework == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.Framework must not be empty")
+	}
+
+	if config.ResyncPeriod == 0 {
+		return nil, microerror.Maskf(invalidConfigError, "config.ResyncPeriod must not be zero")
+	}
+
+	newOperator := &Operator{
+		// Dependencies.
+		backOff:   config.BackOff,
+		framework: config.Framework,
+		k8sClient: config.K8sClient,
+		logger:    config.Logger,
+
+		// Internals.
+		bootOnce:     sync.Once{},
+		mutex:        sync.Mutex{},
```

Comment:
I am not sure why this is `Mutex` needed.

Reply:
It is not. Lets remove it.

Reply:
Done.